### PR TITLE
Work cancellation and timeout handling

### PR DESCRIPTION
**Pull request by kennethlakin**
This addresses Issue #109 . See the commit comments for more details.

Do note that it's based off of my check_uri-handle-iodata branch, mentioned in PR #105 . (This is probably the source of the branch conflict notification below.) I can rebase it if you like, but the iodata handling change is small and nice to have. :)

### CODE REVIEW

**Comment by elbrujohalcon** [comment]
@kennethlakin you'll have to rebase/merge from master

**Comment by elbrujohalcon** [comment]
@kennethlakin #105 was merged, but this one still needs rebasing

**Comment by kennethlakin** [comment]
Rebased.

I don't know why Elvis is complaining here. ```make elvis``` has no complaints on my machine.

**Comment by kennethlakin** [comment]
Actually, hang on. There are some whitespace changes that I accidentally un-did. Gimmy a few minutes to fix that.

**Comment by kennethlakin** [comment]
Okay. Whitespace issues are fixed. Feel free to merge.

File: src/shotgun.erl [hunk 1]
```diff
@@ -393,12 +420,14 @@ at_rest(timeout, State) ->
             ok = gen_fsm:send_event(self(), Work),
             {next_state, at_rest, NewState}
     end;
-at_rest({get_async, {HandleEvent, AsyncMode}, Args, From},
+at_rest({get_async, {HandleEvent, AsyncMode},
+                   {ReqRef, _, _, _} = Args, From},
```
Comment (jfacorro) [inline src/shotgun.erl]: Please check this line's indentation.

Comment (kennethlakin) [inline src/shotgun.erl]: Okay. That line's indentation now matches the ```gun_response``` clauses of ```wait_response/2```.

**Comment by kennethlakin** [comment]
Okay. The line in question's indentation now matches the gun_response clauses of wait_response/2.

**Comment by kennethlakin** [comment]
It occurred to me that I had inadvertently changed the return value when ```shotgun:request/6``` times out from ```{error, {timeout, _}}``` to ```{error, timeout}```. That's fixed now.

**Comment by kennethlakin** [comment]
Okay, I played around with this some more this weekend and discovered a serious bug. I'll close the PR for now and reissue it when I've pushed the fix.

**Pull request closed without merging**